### PR TITLE
OBS/359: Regulate Logging format  and implement logging and tracing (traceId in MDC) using AoP.

### PR DESCRIPTION
**Pull request by Johnny850807**
# Now, we can trace in the log

Resolve: #359 

The log will have the output as below:

Note that, there are two MDC fields `remoteAddr` and `traceId`  in the log.

The following logs demonstrate the AnswerQuestion usecase.

```properties
academy-service_1     | 2021-06-10 11:36:12.598  TRACE - 172.18.0.2 traceId='58df0c021e926523a5dee44be010af8b' - .j.s.a.i.ControllerInstrumentationFilter : [Incoming HTTP Request] POST http://localhost/api/exams/1/problems/70013/C/students/1/answers Headers: host=localhost user-agent=PostmanRuntime/7.26.8 authorization=Bearer eyJhbGciOiJIUzI1NiJ9.eyJzdHVkZW50SWQiOjEsImlzQWRtaW4iOnRydWUsImV4cCI6MzAyMTA1OTQ1Mzc3OX0.3Fg_t2Gsp-bYo_AQYPxowQvGQ77FXZTeo0-BAalU29A , Body: content-length=327 content-type=multipart/form-data; boundary=--------------------------506417783512162861808619
academy-service_1     | 2021-06-10 11:36:12.614   INFO - 172.18.0.2 traceId='58df0c021e926523a5dee44be010af8b' - t.w.j.a.d.u.exam.AnswerQuestionUseCase   : [Answer the Question] examId=1 problemId=70013 langEnvName=C studentId=1 fileCount=1
academy-service_1     | 2021-06-10 11:36:12.849  TRACE - 172.18.0.2 traceId='58df0c021e926523a5dee44be010af8b' - s.a.i.InstrumentationRetrofitInterceptor : [Outgoing Http Request] POST http://submission-service/api/problems/70013/C/students/1/submissions
submission-service_1  | 2021-06-10 11:36:12.856  TRACE - 172.18.0.4 traceId='58df0c021e926523a5dee44be010af8b' - .j.s.a.i.ControllerInstrumentationFilter : [Incoming HTTP Request] POST http://submission-service/api/problems/70013/C/students/1/submissions Headers: authorization=Bearer eyJhbGciOiJIUzI1NiJ9.eyJzdHVkZW50SWQiOi0xMTIsImlzQWRtaW4iOnRydWUsImV4cCI6MTYyMzMzOTM3Mn0.6pQSljeOuAc05rkxI5QGb3x5sTdULl6OE9UioslrDlk host=submission-service connection=Keep-Alive user-agent=okhttp/3.14.8, Body: content-length=-1 content-type=multipart/form-data; boundary=756de135-feec-4b2a-95b8-37a48b24bb59
submission-service_1  | 2021-06-10 11:36:12.864   INFO - 172.18.0.4 traceId='58df0c021e926523a5dee44be010af8b' - t.w.j.s.d.usecases.SubmitCodeUseCase     : [Submit Code] SubmitCodeRequest(throttle=false, problemId=70013, languageEnvName=C, studentId=1, fileResources=[tw.waterball.judgegirl.commons.models.files.FileResource@2dc49a73], submissionBag={examId=1, broker-additional-destinations=/exams/1})
submission-service_1  | 2021-06-10 11:36:12.868  TRACE - 172.18.0.4 traceId='58df0c021e926523a5dee44be010af8b' - trumentationClientHttpRequestInterceptor : [Outgoing Http Request] GET http://problem-service:80/api/problems/70013
problem-service_1     | 2021-06-10 11:36:12.876  TRACE - 172.18.0.6 traceId='58df0c021e926523a5dee44be010af8b' - .j.s.a.i.ControllerInstrumentationFilter : [Incoming HTTP Request] GET http://problem-service/api/problems/70013 Headers: authorization=Bearer eyJhbGciOiJIUzI1NiJ9.eyJzdHVkZW50SWQiOi0xMTEsImlzQWRtaW4iOnRydWUsImV4cCI6MTYyMzMzOTM3Mn0.X7LzIR_A_vnC1b7IHTfvZIuOi6fNHcBvsAG84dOzhwY user-agent=Java/11.0.9.1 host=problem-service connection=keep-alive, Body: content-length=-1 content-type=null
submission-service_1  | 2021-06-10 11:36:12.997  TRACE - 172.18.0.4 traceId='58df0c021e926523a5dee44be010af8b' - t.w.j.s.d.usecases.SubmitCodeUseCase     : [Submission Saved] submissionId="60c1f92c1924497f598ffbf0"
submission-service_1  | 2021-06-10 11:36:15.131  TRACE - 172.18.0.4 traceId='58df0c021e926523a5dee44be010af8b' - t.w.j.s.d.usecases.SubmitCodeUseCase     : [Judger Deployed] submissionId="60c1f92c1924497f598ffbf0"
submission-service_1  | 2021-06-10 11:36:15.132   INFO - 172.18.0.4 traceId='58df0c021e926523a5dee44be010af8b' - t.w.judgegirl.primitives.EventBus        : [Produce: LiveSubmissionEvent] problemId=70013 languageEnvName=C studentId=1 submissionId=60c1f92c1924497f598ffbf0 submissionTime=1623324972942, with bag: examId="1" broker-additional-destinations="/exams/1"
academy-service_1     | 2021-06-10 11:36:15.136  DEBUG - 172.18.0.2 traceId='58df0c021e926523a5dee44be010af8b' - t.w.j.api.retrofit.BaseRetrofitAPI       : [API Response] Response{protocol=http/1.1, code=202, message=, url=http://submission-service/api/problems/70013/C/students/1/submissions}

```

### CODE REVIEW

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Spring-Boot/Spring-Boot-Commons/src/main/java/tw/waterball/judgegirl/springboot/advices/instrumentation/ControllerInstrumentationFilter.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import static tw.waterball.judgegirl.springboot.advices.instrumentation.TraceIds.generateTraceId;
+
+/**
+ * @author - johnny850807@gmail.com (Waterball)
+ */
+@Slf4j
+@Order(Ordered.HIGHEST_PRECEDENCE)
+@Component
+public class ControllerInstrumentationFilter extends GenericFilterBean {
+    public static final Set<String> INCLUDED_HEADERS = new HashSet<>(asList("authorization", "user-agent", "host", "connection"));
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply Spring-Boot/Spring-Boot-Commons/src/main/java/tw/waterball/judgegirl/springboot/advices/instrumentation/ControllerInstrumentationFilter.java:30]: great suggestion

File: Spring-Boot/Spring-Boot-Commons/src/main/java/tw/waterball/judgegirl/springboot/advices/instrumentation/TraceIds.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    private static final int BYTES_LENGTH = 16;
+    private static final int HEX_LENGTH = 2 * BYTES_LENGTH;
+    private static final String INVALID = "00000000000000000000000000000000";
+    private static final Random RANDOM = new Random();
+
+    private TraceIds() {
+    }
+
+    public static String generateTraceId() {
+        byte[] bytes = new byte[16];
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply Spring-Boot/Spring-Boot-Commons/src/main/java/tw/waterball/judgegirl/springboot/advices/instrumentation/TraceIds.java:36]: cool, you even know how to judge their code
https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/TraceId.java

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/AnswerQuestionUseCase.java [hunk 3]
```diff
@@ -56,16 +56,16 @@ private void onlyExamineeCanAnswerQuestion(Request request) throws ExamineeOnlyO
...
+        if (!exam.isOngoing()) {
+            throw new ExamHasNotBeenStartedOrHasBeenClosedException(exam);
         }
     }
 
     private void answerCountMustNotExceedSubmissionQuota(Request request, Question question) throws NoSubmissionQuotaException {
         int answerCount = examRepository.countAnswersInQuestion(question.getId(), request.studentId);
         if (answerCount >= question.getQuota()) {
-            throw new NoSubmissionQuotaException();
+            throw new NoSubmissionQuotaException(question.getQuota());
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/AnswerQuestionUseCase.java:68]: I think this is fine

File: Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/submission/events/LiveSubmissionEvent.java [hunk 4]
```diff
@@ -1,25 +1,37 @@
...
+public class LiveSubmissionEvent extends Event {
     private final int problemId;
     private final String languageEnvName;
     private final int studentId;
     private final String submissionId;
     private final Date submissionTime;
     private final Bag submissionBag;
 
+    public LiveSubmissionEvent(int problemId, String languageEnvName, int studentId, String submissionId, Date submissionTime, Bag submissionBag) {
+        super("LiveSubmissionEvent");
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/submission/events/LiveSubmissionEvent.java:26]: `getClass` can't be used within the `super` braces. I write `super(LiveSubmissionEvent.class.getSimpleName());` instead

File: Spring-Boot/Spring-Boot-Commons/src/main/java/tw/waterball/judgegirl/springboot/advices/CommonExceptionAdvices.java [hunk 5]
```diff
@@ -37,53 +39,37 @@ public CommonExceptionAdvices() {
     @ExceptionHandler({NotFoundException.class})
     public ResponseEntity<?> handleNotFoundException(Exception err) {
         if (log.isDebugEnabled()) {
-            log.error("[NotFound]", err);
-        } else {
-            log.error("[NotFound] {\"message\": \"{}\"}", err.getMessage());
+            log.debug("[Resource Not Found] {}", err.getMessage());
         }
-        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(err.getMessage());
+        return status(HttpStatus.NOT_FOUND).body(err.getMessage());
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply Spring-Boot/Spring-Boot-Commons/src/main/java/tw/waterball/judgegirl/springboot/advices/CommonExceptionAdvices.java:44]: I think (1)  is okay.

**Review by Wally5077 — APPROVED** [review]
LGTM

**Pull request merged**